### PR TITLE
Add a keycloak module to query keys metadata

### PR DESCRIPTION
**Pull request by thomasbach-dev**
##### SUMMARY

When setting up Nextcloud integration to Keycloak we have to ship the metadata of a realm to Nextcloud. With the current modules it is not possible to extract the metadata. This PR adds a module `keycloak.keycloak_realm_keys_metadata` which makes it possible to query the metadata and pass it on:

```yaml
- name: Fetch Keys metadata
  community.general.keycloak_realm_keys_metadata:
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
    realm: MyCustomRealm
  delegate_to: localhost
  register: keys_metadata

- name: Write the Keycloak keys certificate into a file
  ansible.builtin.copy:
    dest: /tmp/keycloak.cert
    content: |
      {{ keys_metadata['keys_metadata']['keys']
      | selectattr('algorithm', 'equalto', 'RS256')
      | map(attribute='certificate')
      | first
      }}
  delegate_to: localhost
```

The module presented here is already used in production in our Ansible playbooks and is known to work.

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

`keycloak_realm_keys_metadata`

### CODE REVIEW

**Comment by thomasbach-dev** [comment]
I added myself to the list of maintainers in the KeyCloak team to get sanity checks to get green.
I am using Keycloak and Ansible heavily in a current project. So I actually might have more contributions. :)

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (russoz) [inline plugins/module_utils/identity/keycloak/keycloak.py:314]: Typo?
```suggestion
        :return: dict with realm configuration, representation or None if none matching exist
```

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (russoz) [inline plugins/modules/keycloak_realm_keys_metadata.py:18]: 9.1.0 has already been released, this module will go into the next one
```suggestion
version_added: 9.2.0
```

**Review by russoz — COMMENTED** [review]
Hi @thomasbach-dev 
Thanks for your contribution. There is a couple of adjustments to make, other than that it looks  good.

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata.py:35]: For `_info` modules, remove this:
```suggestion
```

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata.py:45]: ... and add this instead:
```suggestion
    - community.general.attributes
    - community.general.attributes.info_module
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please note that every new module must come with tests (unit tests, integration tests, or both).

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (thomasbach-dev) [reply plugins/module_utils/identity/keycloak/keycloak.py:314]: Looks like a typo, right? :) tbh, I just copy pasted this from other descriptions and did not really get what `dict of real` could possibly mean. It makes sense now -- the typo propagated.
I'll have a proper look again what the returning data type is here

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata.py]: Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata.py]: Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2 and a87ba3e2968281d7a97a58efa91930d49561de9e

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata.py]: Adressed in a9de64c1937efd3db2ef7608485987f8b698adb2

File: plugins/modules/keycloak_realm_keys_metadata.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata.py]: Addressed in a9de64c1937efd3db2ef7608485987f8b698adb2

**Comment by thomasbach-dev** [comment]
Sorry for the late reply -- very busy weeks. I incorporated the easy fixes for now. I will properly document the return type of the added method in `keycloak.py` and see how I can test this.

**Comment by thomasbach-dev** [comment]
I added a test in 2019594c5d56650b742912cbf718519cb20e4660 following the testing pattern I found in other `test_keycloak_*` modules.

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (russoz) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:79]:
```suggestion
        - Representation of the realm keys metadata (see
          U(https://www.keycloak.org/docs-api/latest/rest-api/index.html#KeysMetadataRepresentation)).
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (russoz) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:18]: 9.2.0 has been released, moving to next one
```suggestion
version_added: 9.3.0
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (russoz) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:104]: Personally I think this comment is redundant. It does not add much, IMHO.

File: tests/unit/plugins/modules/test_keycloak_realm_keys_metadata_info.py
Comment (russoz) [inline tests/unit/plugins/modules/test_keycloak_realm_keys_metadata_info.py:42]: I think the readability of this code would be improved if this `with``were to stay in only one line. Rename the mock object to something shorter - this symbol name is used only in the very next line, having an explicit name does not add more clarity than the line where it is defined, IMHO.

File: plugins/module_utils/identity/keycloak/keycloak.py
Comment (russoz) [reply plugins/module_utils/identity/keycloak/keycloak.py]: Looks better. :-)

**Review by russoz — COMMENTED** [review]
Hi @thomasbach-dev , thanks for the updates. There is a couple of other minor stuff in there, but overall looks good.

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: Picked up in c6bee82ea1235c29bdf27c10ca1447caa149925b

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: Yes, I agree. I removed the comment in c6bee82ea1235c29bdf27c10ca1447caa149925b

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: Picked up in c6bee82ea1235c29bdf27c10ca1447caa149925b

File: tests/unit/plugins/modules/test_keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply tests/unit/plugins/modules/test_keycloak_realm_keys_metadata_info.py]: Well, I use black to automatically format my code when saving. I had to disable it in this case, as I would not get the line below 80 characters -- which is the default max line length in black. Unless I am missing something, there is no style guideline for these things in this repo, right?

c6bee82ea1235c29bdf27c10ca1447caa149925b contains a version which looks like this:
```python
@contextmanager
def patch_keycloak_api(side_effect):
    # …
    obj = keycloak_realm_keys_metadata_info.KeycloakAPI
    with patch.object(obj, "get_realm_keys_metadata_by_id", side_effect=side_effect) as obj_mocked:
        yield obj_mocked
```

where the long line is 99 characters wide which seems to be accepted style. :)

**Review by russoz — APPROVED** [review]
LGTM

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:85]: https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#avoid-using-latin-phrases
```suggestion
            description: A mapping (that is, a dict) from key algorithms to UUIDs.
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:90]: `type=list` should always come with `elements`:
```suggestion
            type: list
            elements: dict
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:55]: How about adjusting this name a bit...
```suggestion
  register: keycloak_keys_metadata
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:61]: ... so that here it's more clear which of the two `keys_metadata` is the return value of the module? Might be helpful for beginners.
```suggestion
      {{ keycloak_keys_metadata['keys_metadata']['keys']
```

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (felixfontein) [inline plugins/modules/keycloak_realm_keys_metadata_info.py:28]: Is this last paragraph useful for this module? I'm wondering whether we should remove it for the `_info` modules. On the other hand, it doesn't really hurt either...

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: Incorporated in latest fixup commit a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py:86]: Incorporated in latest fixup commit a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: Sounds good to me, also incorporated in a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

File: plugins/modules/keycloak_realm_keys_metadata_info.py
Comment (thomasbach-dev) [reply plugins/modules/keycloak_realm_keys_metadata_info.py]: I removed the paragraph. I think it does not belong here, if it's not relevant for this module. Thanks for pointing this out. Fixed in a8370a3101f7b38fb5c48c64c4ab2e947e481ebe

**Comment by thomasbach-dev** [comment]
So, yet another round: I incorporated the suggestions from @felixfontein -- thanks for the thorough examination! I added these as a fixup commit which I will squash if I get a thumbs up. :)
I also rebased on `main`.

**Comment by russoz** [comment]
Still LGTM

**Comment by felixfontein** [comment]
@thomasbach-dev no need to squash, since we squash on merge anyway :) But if you prefer to squash before merge you can also do that...

**Comment by thomasbach-dev** [comment]
Hmm, okay. So is this ready to get merged? Is there something left to do from my side?

**Comment by felixfontein** [comment]
No, just a last quick look and then merge :)

@thomasbach-dev thanks for your contribution!
@russoz thanks for reviewing!

**Comment by felixfontein** [comment]
FYI @thomasbach-dev you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746

**Pull request merged**
